src/Shared/PackageManagers/ProjectManagerFactory.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Copyright (c) Microsoft Corporation. Licensed under the MIT License.
+
+namespace Microsoft.CST.OpenSource.PackageManagers
+{
+    using Contracts;
+    using Model.PackageActions;
+    using PackageUrl;
+    using System;
+    using System.Collections.Generic;
+    using System.Net.Http;
+
+    public class ProjectManagerFactory
+    {
+        /// <summary>
+        /// The delegate to use when defining the function to create a project manager with.
+        /// The only runtime parameter we need is the destination directory. Everything else can be defined in the constructor.
+        /// </summary>
+        private delegate BaseProjectManager? ConstructProjectManager(string destinationDirectory = ".");
+
+        /// <summary>
+        /// The dictionary of project managers.
+        /// </summary>
+        private readonly Dictionary<string, ConstructProjectManager> _projectManagers;
+
+        /// <summary>
+        /// Initializes a new instance of a <see cref="ProjectManagerFactory"/>.
+        /// </summary>
+        /// <param name="httpClientFactoryParam">The <see cref="IHttpClientFactory"/> to use in the project managers.</param>
+        /// <param name="nugetPackageActions">The <see cref="NuGetPackageActions"/> to use in the <see cref="NuGetProjectManager"/>.</param>
+        public ProjectManagerFactory(IHttpClientFactory? httpClientFactoryParam = null, NuGetPackageActions? nugetPackageActions = null)
```

Comment:
```suggestion
        public ProjectManagerFactory(IHttpClientFactory? httpClientFactoryParam = null, IManagerPackageActions? managerPackageActions = null)
```

Reply:
The issue with the factory now is that it's not exactly DI, but trying to mimic how DI would work. So in that case, we would need one parameter for each project manager that uses a `IManagerPackageActions` in the ProjectManager implementation. So say one gets added for NPM, the `ProjectManagerFactory` constructor would have to be:
`public ProjectManagerFactory(IHttpClientFactory? httpClientFactoryParam = null, NuGetPackageActions? nugetPackageActions = null, NpmPackageActions? npmPackageActions = null)` so that way we can override one, the other, both, or neither.

Reply:
In other words, I don't think we can change it to just `managerPackageActions`, as it will have to be manager specific at least given this new design.

Reply:
Given that the ctor has these tightly coupled arguments I don't think this is a good choice to use for the public API for the reasons you mention. This simply does not scale.

Here's my proposal similar to before, to help avoid some of this reflection:

I think this class should have this ctor

```
public ProjectManagerFactory(Dictionary<string, ConstructProjectManager>? startingManagerSet = null)
{
   // Keep the existing default behavior if you don't provide any managers, otherwise use the set the user provided
    _projectManagers = startingManagerSet ?? GenerateDefaultSet();
...
}
```

```
public Dictionary<string, ConstructProjectManager> GenerateDefaultSet(IHttpClientFactory? factory = null)
{
// Existing code that generates the set below
}
```

It should have these methods that manipulate the dictionary
```
// Replace the manager for the specified purl type.
public bool SetManager(string purlType, ConstructProjectManager constructMethod)
// Replace the dictionary with the provided dictionary
public bool SetManagers(IDictionary<string, ConstructProjectManager> managersToReplace)
// Remove the manager for the specified purl type
public bool RemoveManager(string purlType)
// Remove all entries from the dictionary
public bool ClearManagers()
```